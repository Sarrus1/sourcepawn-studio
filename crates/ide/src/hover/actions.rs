use hir::DefResolution;
use ide_db::RootDatabase;

use super::HoverAction;

pub fn goto_type_action_for_def(db: &RootDatabase, def: DefResolution) -> Option<HoverAction> {
    let res = match def {
        DefResolution::Function(it) => it.type_def(db),
        DefResolution::Methodmap(it) => it.type_def(db),
        DefResolution::Property(it) => it.type_def(db),
        DefResolution::Typedef(it) => it.type_def(db),
        DefResolution::Functag(it) => it.type_def(db),
        DefResolution::Field(it) => it.type_def(db),
        DefResolution::Global(it) => it.type_def(db),
        DefResolution::Local(it) => it.type_def(db),
        DefResolution::Variant(it) => it.type_def(db),
        DefResolution::Macro(_)
        | DefResolution::Enum(_)
        | DefResolution::Typeset(_)
        | DefResolution::EnumStruct(_)
        | DefResolution::Funcenum(_)
        | DefResolution::File(_) => vec![],
    };

    Some(HoverAction::goto_type_from_targets(db, res))
}
