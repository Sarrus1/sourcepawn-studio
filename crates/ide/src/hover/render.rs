use hir::DefResolution;
use ide_db::RootDatabase;
use vfs::FileId;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Render {
    FileId(FileId),
    String(String),
}

impl From<FileId> for Render {
    fn from(file_id: FileId) -> Self {
        Render::FileId(file_id)
    }
}

impl From<String> for Render {
    fn from(string: String) -> Self {
        Render::String(string)
    }
}

pub fn render_def(db: &RootDatabase, def: DefResolution) -> Option<Render> {
    Some(match def {
        DefResolution::Function(it) => it.render(db)?.into(),
        DefResolution::Macro(it) => it.render(db)?.into(),
        DefResolution::EnumStruct(it) => it.render(db)?.into(),
        DefResolution::Methodmap(it) => it.render(db)?.into(),
        DefResolution::Property(it) => it.render(db)?.into(),
        DefResolution::Enum(it) => it.render(db)?.into(),
        DefResolution::Variant(it) => it.render(db)?.into(),
        DefResolution::Field(it) => it.render(db)?.into(),
        DefResolution::Typedef(it) => it.render(db)?.into(),
        DefResolution::Typeset(it) => it.render(db)?.into(),
        DefResolution::Functag(it) => it.render(db)?.into(),
        DefResolution::Funcenum(it) => it.render(db)?.into(),
        DefResolution::Global(it) => it.render(db)?.into(),
        DefResolution::Local(it) => it.render(db)?.into(),
        DefResolution::File(it) => it.file_id().into(),
    })
}
