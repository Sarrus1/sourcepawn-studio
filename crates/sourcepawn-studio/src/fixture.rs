use anyhow::Result;
use crossbeam::channel::Receiver;
use itertools::Itertools;
use lsp_server::{Connection, Response};
use lsp_types::{
    notification::{DidOpenTextDocument, Exit, Initialized},
    request::{Completion, Initialize, ResolveCompletionItem, Shutdown},
    ClientCapabilities, CompletionContext, CompletionItem, CompletionItemKind, CompletionParams,
    CompletionResponse, CompletionTriggerKind, DidOpenTextDocumentParams, InitializeParams,
    InitializedParams, Location, LocationLink, Position, Range, SignatureHelp,
    TextDocumentIdentifier, TextDocumentItem, TextDocumentPositionParams, Url, WorkspaceFolder,
};
use std::{
    env,
    fs::File,
    io,
    path::{Path, PathBuf},
    str::FromStr,
    sync::Once,
    thread::JoinHandle,
    time::Duration,
};
use tempfile::{tempdir, TempDir};
use zip::ZipArchive;

use crate::{config::ConfigData, lsp};

use super::{GlobalState, LspClient};

#[derive(Debug)]
pub enum InternalMessage {
    OptionsRequested,
}

#[derive(Debug)]
pub struct Fixture {
    pub documents: Vec<Document>,
}

impl Fixture {
    pub fn parse(input: &str) -> Fixture {
        let mut documents = Vec::new();

        let mut start = 0;
        if !input.is_empty() {
            for end in input
                .match_indices("%!")
                .skip(1)
                .map(|(i, _)| i)
                .chain(std::iter::once(input.len()))
            {
                documents.push(Document::parse(&input[start..end]));
                start = end;
            }
        }

        Self { documents }
    }

    pub fn write_files(&self, dir: &Path) {
        for document in &self.documents {
            let text = String::from(&document.text);
            let path = dir.join(&document.path);
            std::fs::create_dir_all(path.parent().unwrap()).unwrap();
            std::fs::write(&path, &text).unwrap();
        }
    }

    pub fn setup(&self, client: &LspClient, dir: &Path) {
        for document in &self.documents {
            let text = String::from(&document.text);
            let path = dir.join(&document.path);

            let uri = Url::from_file_path(&path).unwrap();
            let language_id = "sourcepawn".to_string();

            client
                .send_notification::<DidOpenTextDocument>(DidOpenTextDocumentParams {
                    text_document: TextDocumentItem::new(uri, language_id, 0, text),
                })
                .unwrap();
        }
    }
}

#[derive(Debug)]
pub struct Document {
    pub path: PathBuf,
    pub text: String,
    pub cursor: Option<Position>,
    pub ranges: Vec<Range>,
}

impl Document {
    pub fn parse(input: &str) -> Self {
        let mut lines = Vec::new();

        let (path, input) = input
            .trim()
            .strip_prefix("%! ")
            .map(|input| input.split_once('\n').unwrap_or((input, "")))
            .unwrap();

        let mut ranges = Vec::new();
        let mut cursor = None;

        for line in input.lines() {
            if line.chars().all(|c| matches!(c, ' ' | '^' | '|' | '!')) && !line.is_empty() {
                let index = (lines.len() - 1) as u32;

                cursor = cursor.or_else(|| {
                    let character = line.find('|')?;
                    Some(Position::new(index, character as u32))
                });

                if let Some(start) = line.find('!') {
                    let position = Position::new(index, start as u32);
                    ranges.push(Range::new(position, position));
                }

                if let Some(start) = line.find('^') {
                    let end = line.rfind('^').unwrap() + 1;
                    ranges.push(Range::new(
                        Position::new(index, start as u32),
                        Position::new(index, end as u32),
                    ));
                }
            } else {
                lines.push(line);
            }
        }

        Self {
            path: PathBuf::from(path),
            text: lines.join("\n"),
            cursor,
            ranges,
        }
    }
}

static LOGGER: Once = Once::new();

#[derive(Debug)]
pub struct TestBed {
    fixture: Fixture,
    locations: Vec<Location>,
    _temp_dir: TempDir,
    temp_dir_path: PathBuf,
    #[allow(unused)]
    temp_sm_dir_path: PathBuf,
    pub internal_rx: Receiver<InternalMessage>,
    client: LspClient,
    client_thread: Option<JoinHandle<()>>,
    server_thread: Option<JoinHandle<()>>,
}

impl Drop for TestBed {
    fn drop(&mut self) {
        log::trace!("Dropping server.");
        let _ = self.client.send_request::<Shutdown>(());
        let _ = self.client.send_notification::<Exit>(());
        self.client_thread.take().unwrap().join().unwrap();
        self.server_thread.take().unwrap().join().unwrap();
    }
}

impl TestBed {
    pub fn new(fixture: &str, add_sourcemod: bool) -> Result<Self> {
        LOGGER.call_once(|| {
            if option_env!("TEST_LOG") == Some("1") {
                fern::Dispatch::new()
                    .level(log::LevelFilter::Trace)
                    .chain(std::io::stderr())
                    .apply()
                    .unwrap()
            }
        });

        let fixture = Fixture::parse(fixture);

        let temp_dir = tempdir()?;
        let temp_dir_path = dunce::canonicalize(temp_dir.path())?;

        let temp_sm_dir = tempdir()?;
        let temp_sm_dir_path = dunce::canonicalize(temp_sm_dir.path())?;
        let temp_sm_dir_path_ = temp_sm_dir_path.clone(); // Copy the value to be able to move it into the closure

        let locations: Vec<Location> = fixture
            .documents
            .iter()
            .flat_map(|document| {
                let uri = Url::from_file_path(temp_dir_path.join(&document.path)).unwrap();
                document
                    .ranges
                    .iter()
                    .map(move |range| Location::new(uri.clone(), *range))
            })
            .collect();

        let (server_conn, client_conn) = Connection::memory();
        let (internal_tx, internal_rx) = crossbeam::channel::unbounded();

        let client = LspClient::new(client_conn.sender);

        let server_thread =
            std::thread::spawn(move || GlobalState::new(server_conn, false).run().unwrap());
        let client_thread = {
            let client = client.clone();
            std::thread::spawn(move || {
                let destination = temp_sm_dir_path_;
                for message in &client_conn.receiver {
                    match message {
                        lsp_server::Message::Request(request) => {
                            if request.method == "workspace/configuration" {
                                #[allow(unused_mut)]
                                let mut config = ConfigData::default();
                                if add_sourcemod {
                                    let mut current_dir = env::current_dir().unwrap();
                                    // The env depends on if we use the debugger or not.
                                    if current_dir.ends_with("sourcepawn-vscode") {
                                        current_dir = current_dir.join("crates/sourcepawn-studio/");
                                    }
                                    let sourcemod_path =
                                        current_dir.join("test_data/sourcemod.zip");
                                    unzip_file(&sourcemod_path, &destination).unwrap();
                                    #[cfg(test)]
                                    config
                                        .includeDirectories
                                        .push(destination.clone().join("include/"));
                                }
                                client
                                    .send_response(Response::new_ok(request.id, vec![config]))
                                    .unwrap();
                                std::thread::sleep(Duration::from_millis(100));
                                internal_tx.send(InternalMessage::OptionsRequested).unwrap()
                            } else {
                                client
                                    .send_error(
                                        request.id,
                                        lsp_server::ErrorCode::MethodNotFound,
                                        "Method not found".into(),
                                    )
                                    .unwrap();
                            }
                        }
                        lsp_server::Message::Response(response) => {
                            client.recv_response(response).unwrap();
                        }
                        lsp_server::Message::Notification(_) => {}
                    }
                }
            })
        };
        fixture.write_files(&temp_dir_path);

        Ok(TestBed {
            fixture,
            locations,
            _temp_dir: temp_dir,
            temp_dir_path,
            temp_sm_dir_path,
            client,
            internal_rx,
            client_thread: Some(client_thread),
            server_thread: Some(server_thread),
        })
    }

    pub fn initialize(&self, capabilities: ClientCapabilities) -> Result<()> {
        self.client
            .send_request::<Initialize>(InitializeParams {
                capabilities,
                workspace_folders: Some(vec![WorkspaceFolder {
                    uri: Url::from_file_path(self.temp_dir_path.clone()).unwrap(),
                    name: "test".to_string(),
                }]),
                root_uri: Some(Url::from_file_path(self.temp_dir_path.clone()).unwrap()),
                ..Default::default()
            })
            .unwrap();
        self.client
            .send_notification::<Initialized>(InitializedParams {})
            .unwrap();

        self.fixture.setup(&self.client, self.directory());

        Ok(())
    }

    pub fn client(&self) -> &LspClient {
        &self.client
    }

    pub fn cursor(&self) -> Option<TextDocumentPositionParams> {
        let (document, cursor) = self
            .fixture
            .documents
            .iter()
            .find_map(|document| document.cursor.map(|cursor| (document, cursor)))?;

        let uri = Url::from_file_path(self.temp_dir_path.join(&document.path)).unwrap();
        let id = TextDocumentIdentifier::new(uri);
        Some(TextDocumentPositionParams::new(id, cursor))
    }

    #[allow(unused)]
    pub fn locations(&self) -> &[Location] {
        &self.locations
    }

    pub fn directory(&self) -> &Path {
        &self.temp_dir_path
    }

    #[allow(unused)]
    pub fn documents(&self) -> &[Document] {
        &self.fixture.documents
    }

    /// Remove the tempdir path from the uri, so that the tests are not dependent
    /// on the tempdir.
    pub fn anonymize_uri(&self, uri: &mut Url) {
        let mut target_path = uri.to_file_path().unwrap();
        target_path = target_path
            .strip_prefix(&self.temp_dir_path)
            .unwrap()
            .to_path_buf();
        uri.set_path(&target_path.to_string_lossy());
    }
}

pub fn goto_definition(fixture: &str) -> Vec<LocationLink> {
    let test_bed = TestBed::new(fixture, true).unwrap();
    test_bed
        .initialize(
            serde_json::from_value(serde_json::json!({
                "textDocument": {
                    "definition": {
                        "linkSupport": true
                    }
                },
                "workspace": {
                    "configuration": true,
                    "workspace_folders": true
                }
            }))
            .unwrap(),
        )
        .unwrap();
    let text_document_position = test_bed.cursor().unwrap();
    let params = lsp_types::request::GotoTypeDefinitionParams {
        text_document_position_params: text_document_position,
        work_done_progress_params: Default::default(),
        partial_result_params: Default::default(),
    };
    let mut locations = match test_bed
        .client()
        .send_request::<lsp_types::request::GotoDefinition>(params)
        .unwrap()
    {
        Some(lsp_types::GotoDefinitionResponse::Link(locations)) => locations,
        _ => unreachable!("Expected a link response."),
    };

    locations.iter_mut().for_each(|location| {
        test_bed.anonymize_uri(&mut location.target_uri);
    });

    locations
}

pub fn complete(fixture: &str, trigger_character: Option<String>) -> Vec<CompletionItem> {
    let test_bed = TestBed::new(fixture, true).unwrap();
    test_bed
        .initialize(
            serde_json::from_value(serde_json::json!({
                "textDocument": {
                    "completion": {
                        "completionItem": {
                            "documentationFormat": ["plaintext", "markdown"]
                        }
                    }
                },
                "workspace": {
                    "configuration": true,
                    "workspace_folders": true
                }
            }))
            .unwrap(),
        )
        .unwrap();
    let text_document_position = test_bed.cursor().unwrap();
    let mut items = match test_bed
        .client()
        .send_request::<Completion>(CompletionParams {
            text_document_position,
            partial_result_params: Default::default(),
            work_done_progress_params: Default::default(),
            context: Some(CompletionContext {
                trigger_kind: CompletionTriggerKind::TRIGGER_CHARACTER,
                trigger_character,
            }),
        })
        .unwrap()
    {
        Some(CompletionResponse::Array(items)) => items,
        Some(CompletionResponse::List(list)) => list.items,
        None => Vec::new(),
    };

    items = items
        .into_iter()
        .map(|item| match item.data {
            Some(_) => {
                let mut item = test_bed
                    .client()
                    .send_request::<ResolveCompletionItem>(item)
                    .unwrap();

                item.data = None;
                item
            }
            None => item,
        })
        .sorted_by(|item1, item2| item1.label.cmp(&item2.label))
        .collect();

    // The results of include completions will always change because the paths of tempdir changes.
    for item in &mut items {
        if matches!(
            item.kind,
            Some(CompletionItemKind::FILE) | Some(CompletionItemKind::FOLDER)
        ) {
            item.detail = item.detail.as_mut().map(|detail| {
                // Account for windows paths
                detail.replace('\\', "/")
            });
        }
    }

    items
}

pub fn signature_help(fixture: &str) -> SignatureHelp {
    let test_bed = TestBed::new(fixture, true).unwrap();
    test_bed
        .initialize(
            serde_json::from_value(serde_json::json!({
                "textDocument": {
                    "signatureHelpProvider": {}
                },
                "workspace": {
                    "configuration": true,
                    "workspace_folders": true
                }
            }))
            .unwrap(),
        )
        .unwrap();
    let text_document_position = test_bed.cursor().unwrap();
    let params = lsp_types::SignatureHelpParams {
        text_document_position_params: text_document_position,
        work_done_progress_params: Default::default(),
        context: None,
    };

    test_bed
        .client()
        .send_request::<lsp_types::request::SignatureHelpRequest>(params)
        .unwrap()
        .unwrap()
}

pub fn unzip_file(zip_file_path: &Path, destination: &Path) -> Result<(), io::Error> {
    let file = File::open(zip_file_path)?;
    let mut archive = ZipArchive::new(file)?;

    for i in 0..archive.len() {
        let mut file = archive.by_index(i)?;
        let dest_path = destination.join(file.name());

        if file.is_dir() {
            std::fs::create_dir_all(&dest_path)?;
        } else {
            if let Some(parent_dir) = std::path::Path::new(&dest_path).parent() {
                std::fs::create_dir_all(parent_dir)?;
            }
            let mut dest_file = File::create(&dest_path)?;
            io::copy(&mut file, &mut dest_file)?;
        }
    }

    Ok(())
}

pub fn hover(fixture: &str) -> lsp::ext::Hover {
    let test_bed = TestBed::new(fixture, true).unwrap();
    test_bed
        .initialize(
            serde_json::from_value(serde_json::json!({
                "textDocument": {
                    "hover": {
                        "contentFormat": [
                            "plaintext",
                            "markdown"
                        ]
                    }
                },
                "workspace": {
                    "configuration": true,
                    "workspace_folders": true
                },
                "experimental": {
                    "hoverActions": true,
                    "commands": {
                        "commands": ["sourcepawn-vscode.gotoLocation"],
                    },
                }
            }))
            .unwrap(),
        )
        .unwrap();
    let text_document_position = test_bed.cursor().unwrap();
    let params = lsp_types::HoverParams {
        text_document_position_params: text_document_position,
        work_done_progress_params: Default::default(),
    };

    let mut res = test_bed
        .client()
        .send_request::<lsp::ext::HoverRequest>(params)
        .unwrap()
        .expect("Expected a hover response.");

    for action in &mut res.actions {
        for command in &mut action.commands {
            if let Some(args) = command.command.arguments.as_mut() {
                for arg in args.iter_mut() {
                    if let Some(arg) = arg.as_object_mut() {
                        if let Some(value) = arg.get_mut("uri") {
                            let mut uri = Url::from_str(value.as_str().unwrap()).unwrap();
                            test_bed.anonymize_uri(&mut uri);
                            *value = serde_json::Value::String(uri.to_string());
                        }
                    }
                }
            }
        }
    }

    res
}
