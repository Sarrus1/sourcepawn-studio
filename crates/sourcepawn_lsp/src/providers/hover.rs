use lsp_types::{Hover, HoverParams};

use super::FeatureRequest;

pub mod description;

pub fn provide_hover(request: FeatureRequest<HoverParams>) -> Option<Hover> {
    let items = &request.store.get_items_from_position(
        request.params.text_document_position_params.position,
        request
            .params
            .text_document_position_params
            .text_document
            .uri
            .clone(),
    );
    if items.is_empty() {
        return None;
    }
    let hover = items[0].read().unwrap().to_hover(&request.params);

    hover
}
