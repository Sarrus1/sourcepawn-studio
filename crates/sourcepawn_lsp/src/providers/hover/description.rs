use lazy_static::lazy_static;
use regex::Regex;

#[derive(Debug, Clone, Default)]
pub struct Description {
    pub text: String,
    pub deprecated: Option<String>,
}

impl Description {
    pub fn to_md(&self) -> String {
        lazy_static! {
            static ref RE1: Regex = Regex::new(r"^\*<").unwrap();
            static ref RE2: Regex = Regex::new(r"\*\s*\r?\n\s*\*").unwrap();
            static ref RE3: Regex = Regex::new(r"\r?\n\s*\*").unwrap();
            static ref RE4: Regex = Regex::new(r"^\*").unwrap();
            static ref RE5: Regex = Regex::new(r"<").unwrap();
            static ref RE6: Regex = Regex::new(r">").unwrap();
            static ref RE7: Regex = Regex::new(r"\s*(@[A-Za-z]+)\s+").unwrap();
            static ref RE8: Regex = Regex::new(r"(_@param_) ([A-Za-z0-9_.]+)\s*").unwrap();
            static ref RE9: Regex = Regex::new(r"(\w+\([A-Za-z0-9_ :]*\))").unwrap();
        }
        let text = RE1.replace_all(&self.text, "").into_owned();
        let text = RE2.replace_all(&text, "\n").into_owned();
        let text = RE3.replace_all(&text, "").into_owned();
        let text = RE4.replace_all(&text, "").into_owned();
        let text = RE5.replace_all(&text, "\\<").into_owned();
        let text = RE6.replace_all(&text, "\\>").into_owned();

        let text = RE7.replace_all(&text, "\n\n_${1}_ ").into_owned();
        let text = RE8.replace_all(&text, "${1} `${2}` â€” >").into_owned();
        let text = RE9.replace_all(&text, "`${1}`").into_owned();
        text.replace("DEPRECATED", "\n\n**DEPRECATED**")
    }
}
