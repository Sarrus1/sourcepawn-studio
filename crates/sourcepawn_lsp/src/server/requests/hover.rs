use crate::utils;
use std::sync::Arc;

use lsp_server::RequestId;
use lsp_types::HoverParams;

use crate::{providers, Server};

impl Server {
    pub(super) fn hover(&mut self, id: RequestId, mut params: HoverParams) -> anyhow::Result<()> {
        utils::normalize_uri(&mut params.text_document_position_params.text_document.uri);
        let uri = Arc::new(
            params
                .text_document_position_params
                .text_document
                .uri
                .clone(),
        );
        let _ = self.read_unscanned_document(uri.clone());

        self.handle_feature_request(id, params, uri, providers::hover::provide_hover)?;

        Ok(())
    }
}
