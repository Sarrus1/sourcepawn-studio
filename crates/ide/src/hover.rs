mod actions;
mod render;

use std::panic::AssertUnwindSafe;

use hir::{DefResolution, HasSource, Semantics};
use ide_db::{Documentation, RootDatabase};
use itertools::Itertools;
use preprocessor::{db::PreprocDatabase, PreprocessingResult};
use smol_str::ToSmolStr;
use syntax::utils::ts_range_to_text_range;
use vfs::FileId;

use crate::{
    events::{event_hover, event_name},
    goto_definition::find_inner_name_range,
    markup::Markup,
    FilePosition, NavigationTarget, RangeInfo,
};

use self::actions::goto_type_action_for_def;
pub(crate) use render::{render_def, Render};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct HoverConfig {
    pub links_in_hover: bool,
    pub documentation: bool,
    pub keywords: bool,
    pub format: HoverDocFormat,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum HoverDocFormat {
    Markdown,
    PlainText,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum HoverAction {
    // Runnable(Runnable),
    Implementation(FilePosition),
    Reference(FilePosition),
    GoToType(Vec<HoverGotoTypeData>),
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct HoverGotoTypeData {
    pub mod_path: String,
    pub nav: NavigationTarget,
}

impl HoverAction {
    fn goto_type_from_targets(db: &RootDatabase, targets: Vec<DefResolution>) -> Self {
        let mut targets = targets
            .into_iter()
            .filter_map(|def| {
                let sema = Semantics::new(db);
                let file_id = def.file_id(db);
                let source_tree = sema.parse(file_id);
                let name = def.name(db).map(|it| it.to_smolstr()).unwrap_or_default();
                let def_node = def.source(db, &source_tree)?.value;

                let name_range = find_inner_name_range(&def_node);

                let target_preprocessing_results = sema.preprocess_file(file_id);
                Some(HoverGotoTypeData {
                    mod_path: Default::default(),
                    nav: NavigationTarget {
                        name,
                        file_id,
                        full_range: target_preprocessing_results
                            .source_map()
                            .closest_u_range_always(ts_range_to_text_range(&def_node.range())),
                        focus_range: target_preprocessing_results
                            .source_map()
                            .closest_u_range_always(name_range)
                            .into(),
                    },
                })
            })
            .collect_vec();
        targets.dedup();
        HoverAction::GoToType(targets)
    }
}

/// Contains the results when hovering over an item
#[derive(Debug, Default)]
pub struct HoverResult {
    pub markup: Markup,
    pub actions: Vec<HoverAction>,
}

pub(crate) fn hover(
    db: &RootDatabase,
    mut fpos: FilePosition,
    config: &HoverConfig,
    file_id_to_url: AssertUnwindSafe<&dyn Fn(FileId) -> Option<String>>,
    events_game_name: Option<&str>,
) -> Option<RangeInfo<HoverResult>> {
    let sema = &Semantics::new(db);
    let preprocessing_results = sema.preprocess_file(fpos.file_id);
    let tree = sema.parse(fpos.file_id);
    let root_node = tree.root_node();
    if let Some(hover) = find_macro_hover(&preprocessing_results, sema, &fpos) {
        return Some(hover);
    }
    fpos.offset = preprocessing_results
        .source_map()
        .closest_s_position_always(fpos.offset);

    let node =
        root_node.descendant_for_byte_range(fpos.raw_offset_usize(), fpos.raw_offset_usize())?;

    if let Some(name) = event_name(&node, &preprocessing_results.preprocessed_text()) {
        return event_hover(
            events_game_name,
            &name,
            &node,
            preprocessing_results.source_map(),
        );
    }

    let def = sema.find_def(fpos.file_id, &node)?;
    let u_range = preprocessing_results
        .source_map()
        .closest_u_range_always(ts_range_to_text_range(&node.range()));

    let file_id = def.file_id(db);
    let source_tree = sema.parse(file_id);
    let text = db.preprocessed_text(file_id);
    let render = render::render_def(db, def.clone())?;
    let mut actions = [goto_type_action_for_def(db, def.clone())]
        .into_iter()
        .flatten()
        .collect_vec();
    actions.dedup();
    let def_node = def.source(db, &source_tree)?.value;

    let markup = match render {
        Render::FileId(file_id) => Markup::from(file_id_to_url(file_id).unwrap_or_default()),
        Render::String(render) => Markup::fenced_block(render),
    };

    if !config.documentation {
        let res = HoverResult { markup, actions };
        return Some(RangeInfo::new(u_range, res));
    }
    if let Some(docs) = Documentation::from_node(def_node, text.as_bytes()) {
        let res = HoverResult {
            markup: Markup::from(format!(
                "{}\n\n---\n\n{}",
                markup,
                Markup::from(docs.to_markdown()),
            )),
            actions,
        };
        return Some(RangeInfo::new(u_range, res));
    }
    let res = HoverResult { markup, actions };
    Some(RangeInfo::new(u_range, res))
}

fn find_macro_hover(
    preprocessing_results: &PreprocessingResult,
    sema: &Semantics<RootDatabase>,
    fpos: &FilePosition,
) -> Option<RangeInfo<HoverResult>> {
    let (offset, def) = sema.find_macro_def(fpos)?;
    let preprocessed_text = preprocessing_results.preprocessed_text();
    let file_id = def.file_id(sema.db);
    let source_tree = sema.parse(file_id);
    let def_node = def.source(sema.db, &source_tree)?.value;
    let source = sema.db.preprocessed_text(file_id);
    let source_text = def_node.utf8_text(source.as_bytes()).ok()?;

    let start: u32 = offset.expanded_range().start().into();
    let end: u32 = offset.expanded_range().end().into();
    let slc = start as usize..end as usize;
    // The preprocessed file might be shorter than the original file
    let hover_text = preprocessed_text
        .get(slc)
        .map(String::from)
        .unwrap_or_default();

    let markup = Markup::from(format!(
        "{}\nExpands to:\n{}",
        Markup::fenced_block(source_text),
        Markup::fenced_block(hover_text.trim())
    ));

    let res = if let Some(docs) = Documentation::from_node(def_node, source.as_bytes()) {
        HoverResult {
            markup: Markup::from(format!(
                "{}\n\n---\n\n{}",
                markup,
                Markup::from(docs.to_markdown()),
            )),
            actions: vec![],
        }
    } else {
        HoverResult {
            markup,
            actions: vec![],
        }
    };

    Some(RangeInfo::new(offset.name_range(), res))
}
