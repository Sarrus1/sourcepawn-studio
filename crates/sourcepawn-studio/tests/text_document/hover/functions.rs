use insta::assert_json_snapshot;
use sourcepawn_studio::fixture::hover;

#[test]
fn function_1() {
    assert_json_snapshot!(hover(
        r#"
%! main.sp
/**
 * The foo function.
 */
void foo(){}
      |
      ^
"#,
    ));
}

#[test]
fn function_2() {
    assert_json_snapshot!(hover(
        r#"
%! main.sp
/**
 * The foo function.
 */
#pragma deprecated Use bar instead.
void foo(){}
      |
      ^
"#,
    ));
}

#[test]
fn function_3() {
    assert_json_snapshot!(hover(
        r#"
%! main.sp
/**
 * The foo function.
 * @param bar The bar parameter.
 *            It is useful.
 * @param baz The baz parameter.
 */
void foo(int bar, int baz){}
              |
              ^
"#,
    ));
}

#[test]
fn function_4() {
    assert_json_snapshot!(hover(
        r#"
%! main.sp
/**
 * The foo function.
 * @param bar The bar parameter.
 *            It is useful.
 * @param baz The baz parameter.
 */
void foo(int bar,
      |
      ^
         int baz){}
"#,
    ));
}

#[test]
fn function_5() {
    assert_json_snapshot!(hover(
        r#"
%! main.sp
methodmap Foo {}
void bar(Foo foo){}
              |
              ^
"#,
    ));
}
