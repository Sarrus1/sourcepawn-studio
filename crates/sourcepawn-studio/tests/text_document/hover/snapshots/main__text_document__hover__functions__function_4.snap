---
source: crates/sourcepawn-studio/tests/text_document/hover/functions.rs
assertion_line: 55
expression: "hover(r#\"\n%! main.sp\n/**\n * The foo function.\n * @param bar The bar parameter.\n *            It is useful.\n * @param baz The baz parameter.\n */\nvoid foo(int bar,\n      |\n      ^\n         int baz){}\n\"#)"
---
{
  "contents": {
    "kind": "markdown",
    "value": "```sourcepawn\nvoid foo(int bar,\n         int baz)\n```\n\n---\n\nThe foo function.\n\n_@param_ `bar` â€” >The bar parameter.\nIt is useful.\n\n_@param_ `baz` â€” >The baz parameter."
  },
  "range": {
    "start": {
      "line": 6,
      "character": 5
    },
    "end": {
      "line": 6,
      "character": 8
    }
  },
  "actions": []
}
